### PR TITLE
Enumerable.Sum Tests

### PR DESCRIPTION
**Pull request by MarcinJuraszek**
Unit tests for `Enumerable.Sum()` and `Enumerable.Sum(selector)`, part of https://github.com/dotnet/corefx/issues/1143.

### CODE REVIEW

File: src/System.Linq/src/System/Linq/Enumerable.Sum.cs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+        public static int? Sum(this IEnumerable<int?> source)
+        {
+            if (source == null) throw Error.ArgumentNull("source");
+            int sum = 0;
+            checked
+            {
+                foreach (int? v in source)
+                {
+                    if (v != null) sum += v.GetValueOrDefault();
```
Comment (JohanLarsson) [inline src/System.Linq/src/System/Linq/Enumerable.Sum.cs]: Why nullcheck here?

Comment (spzSource) [inline src/System.Linq/src/System/Linq/Enumerable.Sum.cs]: Agree. This is redundant code because GetValueOrDefault is used.

Comment (MarcinJuraszek) [inline src/System.Linq/src/System/Linq/Enumerable.Sum.cs]: I didn't change that. Just extracted from `Enumerable.cs` to `Enumerable.Sum.cs`. But agree, we might want to remove the `if` statement as part of the change.

File: src/System.Linq/src/System.Linq.csproj [hunk 2]
```diff
@@ -23,6 +23,7 @@
       <Link>Common\System\NotImplemented.cs</Link>
     </Compile>
     <Compile Include="System\Linq\Enumerable.cs" />
+    <Compile Include="System\Linq\Enumerable.Sum.cs" />
```
Comment (stephentoub) [inline src/System.Linq/src/System.Linq.csproj]: Why the change to separate out just the Sum functions as part of a PR focused on tests?  Seems like if we want to split this file up, it should be a separate discussion/refactoring, at which point we'd do it for all of the members and not just for Sum.

**Comment by stephentoub** [comment]
Other than my one comment about the source being moved, LGTM.

cc: @VSadov

File: src/System.Linq/src/System/Linq/Enumerable.Sum.cs (hunk 1 above)
Comment (VSadov) [inline src/System.Linq/src/System/Linq/Enumerable.Sum.cs]: Just use GetValueOrDefault, since we know we are dealing with an int. Add a comment though that we imply an null check here, but not doing it since the result is the same anyways.

File: src/System.Linq/src/System.Linq.csproj (hunk 2 above)
Comment (VSadov) [inline src/System.Linq/src/System.Linq.csproj]: Yes, if we want to split the file, we should come up with a clear guidance/rule on how to do that and do that considering all operators, not just Sum.
It should be done as a separate change.

File: src/System.Linq/src/System/Linq/Enumerable.Sum.cs (hunk 1 above)
Comment (VSadov) [inline src/System.Linq/src/System/Linq/Enumerable.Sum.cs]: This applies to all "primitive?" cases. Perhaps it is better to do as a separate change and apply to all the cases. There could be other operators using this pattern, and such... 
It would be easier to concentrate on one kind of change.

**Comment by VSadov** [comment]
Same as @stephentoub - nice change, but refactoring into separate files should be done in a separate change.

File: src/System.Linq/src/System/Linq/Enumerable.Sum.cs (hunk 1 above)
Comment (MarcinJuraszek) [inline src/System.Linq/src/System/Linq/Enumerable.Sum.cs]: I agree that it should be fixed as separate change.

File: src/System.Linq/src/System.Linq.csproj (hunk 2 above)
Comment (MarcinJuraszek) [inline src/System.Linq/src/System.Linq.csproj]: OK, reverting that change.

**Comment by MarcinJuraszek** [comment]
Updated the PR to address comments.

**Comment by stephentoub** [comment]
Thanks, @MarcinJuraszek!
